src/merge.jl

```diff
@@ -0,0 +1,236 @@
+using DataStructures # MutableBinaryHeap not found without this line. Why?
+
+# There's a module-level conflict with RegionTrees.vertices. 
+# Is there a better way to resolve it?
+import MetaGraphs
+vertices = MetaGraphs.vertices
+
+"""
+    seg_img = merge(seg, threshold)
+
+Merges segments in a [`SegmentedImage`](@ref) by building a region adjacency
+graph (RAG) and merging segments connected by edges with weight less than 
+`threshold`.
+
+# Arguments:
+* `seg`         : SegmentedImage to be merged.
+* `threshold`   : Upper bound of the adjacent segment color difference to 
+                  consider merging segments.
+
+# Citation:
+Vighnesh Birodkar
+"Hierarchical merging of region adjacency graphs"
+https://vcansimplify.wordpress.com/2014/08/17/hierarchical-merging-of-region-adjacency-graphs/
+"""
+function merge(seg::SegmentedImage, threshold::Number)::SegmentedImage
+    g = seg_to_graph(seg)
+
+    # Populate a heap of all the edges, and a Bool indicating whether the edge
+    # is valid. All edges are initially valid. The reason for this is that heap
+    # removal would be expensive, so instead, we invalidate the edge entry in the
+    # heap. 
+    function weight(t::Tuple{Edge{Int64}, Bool})::Real
+        return has_prop(g, t[1], :weight) ?  get_prop(g, t[1], :weight) : 0
+    end
+    
+    edge_heap = MutableBinaryHeap{Tuple{Edge{Int64}, Bool}}(Base.By(weight), [])
+    sizehint!(edge_heap, 3 * length(edges(g)))  # Overkill, or not enough?
+    for e in edges(g)
+        handle = push!(edge_heap, (e, true))
+        set_prop!(g, e, :handle, handle)
+    end
+
+    # Merge all edges less than threshold
+    while !isempty(edge_heap) && weight(first(edge_heap)) < threshold
+        e, valid = pop!(edge_heap)
+        if valid
+            # Invalidate all edges touching this edge.
+            invalidate_neighbors(g, edge_heap, e)
+
+            # Merge the two nodes into one (keep e.dst, obsolete e.src)
+            merge_node_props!(g, e)
+
+            # Make new edges to the merged node.
+            new_edges = add_neighboring_edges!(g, e)
+
+            # Remove edges to src. 
+            # Don't call rem_vertex!(g, e.src); it would renumber all vertices. 
+            for n in collect(neighbors(g, e.src))
+                rem_edge!(g, e.src, n)
+            end
+
+            # Add new edges to heap.
+            for e in new_edges
+                handle = push!(edge_heap, (e, true))
+                set_prop!(g, e, :handle, handle)
+            end
+        end
+    end
+
+    return resegment(seg, g)
+end
+
+
+"""
+    g = seg_to_graph(seg)
+
+Given a [`SegmentedImage`](@ref), produces a region adjacency [`MetaGraph`](@ref) 
+and stores segment metadata on the vertices. Edge weight is determined by
+color difference.
+
+# Arguments:
+* `seg`         : a [`SegmentedImage`](@ref)
+"""
+function seg_to_graph(seg::SegmentedImage)::MetaGraph
```

Comment:
I predict that someday folks will want to generalize this (e.g., pass a weight function), but this is fine for now.